File: BinaryTree/104maximumDepthofBinaryTree.py
#key: BinaryTree
#Algorithm: recursion

#the key of recursion is the two return, one is the end of trip, the other one is
#the end of the recursion. so in this case, when the recursion ends, return the depth+1


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """ 
        if not root:
            return 0
        else:
            left = self.maxDepth(root.left)
            right = self.maxDepth(root.right)
            return max(left+1,right+1)

File: Tree/104MaximumDepthofBinaryTree.cpp
/** 
 Solution:
 Use recurrsion to solve this problem.
 Base case: root == NULL means arriving a leaf, start counting from here.
 Plus 1 at every parent node. Every paths are counted separately.
 Find the max depth of left and right paths.
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(!root) return 0;
        return max(maxDepth(root->left)+1, maxDepth(root->right)+1);
    }
};